numba/runtime/_nrt_python.c

```diff
@@ -0,0 +1,539 @@
+    int readonly = 0;
+
+    MemInfoObject *miobj = (MemInfoObject*)exporter;
+    MemInfo *mi = miobj->meminfo;
+
+    buf = NRT_MemInfo_data(mi);
+    len = NRT_MemInfo_size(mi);
+    return PyBuffer_FillInfo(view, exporter, buf, len, readonly, flags);
+}
+
+Py_ssize_t MemInfo_rdwrbufferproc(PyObject *self, Py_ssize_t segment,
+                                  void **ptrptr)
+{
+    MemInfoObject *mio = (MemInfoObject *)self;
+    MemInfo *mi = mio->meminfo;
+    if (segment != 0) {
+        PyErr_SetString(PyExc_TypeError, "MemInfo only has 1 segment");
+        return -1;
+    }
+    *ptrptr = NRT_MemInfo_data(mi);
+    return NRT_MemInfo_size(mi);
+}
+
+Py_ssize_t MemInfo_segcountproc(PyObject *self, Py_ssize_t *lenp) {
+    MemInfoObject *mio = (MemInfoObject *)self;
+    MemInfo *mi = mio->meminfo;
+    if (lenp) {
+        *lenp = NRT_MemInfo_size(mi);
+    }
+    return 1;
+}
+
+#if (PY_MAJOR_VERSION < 3)
+static PyBufferProcs MemInfo_bufferProcs = {MemInfo_rdwrbufferproc,
+                                            MemInfo_rdwrbufferproc,
+                                            MemInfo_segcountproc,
+                                            NULL};
+#else
+static PyBufferProcs MemInfo_bufferProcs = {MemInfo_getbuffer, NULL};
+#endif
+
+static
+PyObject*
+MemInfo_acquire(MemInfoObject *self) {
+    NRT_MemInfo_acquire(self->meminfo);
+    Py_RETURN_NONE;
+}
+
+static
+PyObject*
+MemInfo_release(MemInfoObject *self) {
+    NRT_MemInfo_release(self->meminfo, self->defer);
+    Py_RETURN_NONE;
+}
+
+static
+PyObject*
+MemInfo_release_defer(MemInfoObject *self) {
+    NRT_MemInfo_release(self->meminfo, 1);
+    Py_RETURN_NONE;
+}
+
+static
+PyObject*
+MemInfo_set_defer(MemInfoObject *self, PyObject *args) {
+    PyObject *should_defer;
+    int defer;
+    if (!PyArg_ParseTuple(args, "O", &should_defer)) {
+        return NULL;
+    }
+    defer = PyObject_IsTrue(should_defer);
+    if (defer == -1) {
+        return NULL;
+    }
+    self->defer = defer;
+    Py_RETURN_NONE;
+}
+
+
+static
+PyObject*
+MemInfo_get_defer(MemInfoObject *self) {
+    if (self->defer) {
+        Py_RETURN_TRUE;
+    } else {
+        Py_RETURN_FALSE;
+    }
+}
+
+
+static
+PyObject*
+MemInfo_get_data(MemInfoObject *self) {
+    return PyLong_FromVoidPtr(NRT_MemInfo_data(self->meminfo));
+}
+
+static void
+MemInfo_dealloc(MemInfoObject *self)
+{
+    NRT_MemInfo_release(self->meminfo, self->defer);
+    Py_TYPE(self)->tp_free((PyObject*)self);
+}
+
+static PyMethodDef MemInfo_methods[] = {
+    {"acquire", (PyCFunction)MemInfo_acquire, METH_NOARGS,
+     "Increment the reference count"
+    },
+    {"release", (PyCFunction)MemInfo_release, METH_NOARGS,
+     "Decrement the reference count"
+    },
+    {"release_defer", (PyCFunction)MemInfo_release_defer, METH_NOARGS,
+     "Decrement the reference count but defer the destructor regardless"
+     "of the value in defer attribute"
+    },
+    {"set_defer", (PyCFunction)MemInfo_set_defer, METH_VARARGS,
```

Comment:
Just for the record, set_defer / get_defer could have been a "getset" instead (there's an example in numba/npyufunc/_internal.c).